core/com.b2international.snowowl.core/src/com/b2international/snowowl/core/request/resource/BaseMetadataResourceConverter.java

```diff
@@ -220,6 +231,115 @@ protected final void expandDependencyUpgrades(List<R> results) {
+			return;
+		}
+		
+		final List<String> upgradeOfBranches = context().service(ResourceURIPathResolver.class).resolve(context(), upgradeOfURIs);
+		
+		final Map<ResourceURI, String> branchesByUpgradeOf = Maps.newHashMap();
+		Iterator<ResourceURI> uriIterator = upgradeOfURIs.iterator();
+		Iterator<String> branchIterator = upgradeOfBranches.iterator();
+		while (uriIterator.hasNext() && branchIterator.hasNext()) {
+			ResourceURI uri = uriIterator.next();
+			String branch = branchIterator.next();
+			branchesByUpgradeOf.put(uri, branch);
+		}
+
+		for (R result : results) {
+			
+			if (!(result instanceof CodeSystem)) {
+				continue;
+			}
+			
+			CodeSystem cs = (CodeSystem) result;
+			if (cs.getUpgradeOf() == null) {
+				continue;
+			}
+			String upgradeOfCodeSystemBranchPath = branchesByUpgradeOf.get(cs.getUpgradeOf().withoutPath());
+			
+			if (!Strings.isNullOrEmpty(upgradeOfCodeSystemBranchPath)) {
+				RepositoryContext ctx = context().service(RepositoryManager.class).getContext(cs.getToolingId());
+				BaseRevisionBranching branching = ctx.service(BaseRevisionBranching.class);
+				
+				RevisionBranch branch = branching.getBranch(cs.getBranchPath());
+				BranchState branchState = branching.getBranchState(cs.getBranchPath(), upgradeOfCodeSystemBranchPath);
+				BranchInfo mainInfo = new BranchInfo(branch.getPath(), branchState, branch.getBaseTimestamp(), branch.getHeadTimestamp());
+				
+				List<ResourceURI> availableVersions = Lists.newArrayList();
+				List<BranchInfo> versionBranchInfo = Lists.newArrayList();
+
+				if (!cs.getUpgradeOf().isHead()) {
+					long startTimestamp;
+					final String upgradeOfVersionBranch = context().service(ResourceURIPathResolver.class).resolve(context(), List.of(cs.getUpgradeOf())).stream().findFirst().orElse("");
+
+					if (!Strings.isNullOrEmpty(upgradeOfVersionBranch)) {
+						startTimestamp = branching.getBranch(upgradeOfVersionBranch).getBaseTimestamp() + 1;
+					} else {
+						startTimestamp = Long.MIN_VALUE;
+					}
+					
+					ResourceRequests.prepareSearchVersion()
+						.all()
+						.filterByResource(cs.getUpgradeOf().withoutPath())
+						.build()
+						.execute(context())
+						.stream()
+						.filter(csv -> !csv.getVersionResourceURI().isHead())
+						.forEach(csv -> {
+							RevisionBranch versionBranch = branching.getBranch(csv.getBranchPath());
+							
+							if (versionBranch.getParentPath().equals(upgradeOfCodeSystemBranchPath)) {
+								
+								if (versionBranch.getBaseTimestamp() > startTimestamp) {
+									BranchState versionBranchState = branching.getBranchState(cs.getBranchPath(), versionBranch.getPath());
+									if (versionBranchState == BranchState.BEHIND || versionBranchState == BranchState.DIVERGED) {
+										availableVersions.add(csv.getVersionResourceURI());
+									}
+									
+									versionBranchInfo.add(new BranchInfo(branch.getPath(), versionBranchState, versionBranch.getBaseTimestamp(), versionBranch.getHeadTimestamp()));
```

Comment:
I'm not sure if this is safe, or it rather feels weird. When I'm trying to interpret this code I see a BranchInfo object which supposed to collect additional information for one, single branch, right? Yet we are mixing the properties of different branches?

Reply:
Yours is most likely the correct interpretation and `branch` is a local variable that made its way here by accident:

```suggestion
									versionBranchInfo.add(new BranchInfo(versionBranch.getPath(), versionBranchState, versionBranch.getBaseTimestamp(), versionBranch.getHeadTimestamp()));
```